cli/trafficgen.go

```diff
@@ -0,0 +1,276 @@
+			reconnect := uuid.New()
+			conn, err := client.WorkspaceAgentReconnectingPTY(inv.Context(), codersdk.WorkspaceAgentReconnectingPTYOpts{
+				AgentID:   agentID,
+				Reconnect: reconnect,
+				Height:    65535,
+				Width:     65535,
+				Command:   "/bin/sh",
+			})
+			if err != nil {
+				return xerrors.Errorf("connect to workspace: %w", err)
+			}
+
+			defer func() {
+				_ = conn.Close()
+			}()
+
+			// Wrap the conn in a countReadWriter so we can monitor bytes sent/rcvd.
+			crw := countReadWriter{ReadWriter: conn}
+
+			// Set a deadline for stopping the text.
+			start := time.Now()
+			deadlineCtx, cancel := context.WithDeadline(inv.Context(), start.Add(duration))
+			defer cancel()
+
+			// Create a ticker for sending data to the PTY.
+			tick := time.NewTicker(tickInterval)
+			defer tick.Stop()
+
+			// Now we begin writing random data to the pty.
+			writeSize := int(bps / 10)
+			rch := make(chan error)
+			wch := make(chan error)
+
+			// Read forever in the background.
+			go func() {
+				rch <- readContext(deadlineCtx, &crw, writeSize*2)
+				conn.Close()
+				close(rch)
+			}()
+
+			// Write random data to the PTY every tick.
+			go func() {
+				wch <- writeRandomData(deadlineCtx, &crw, writeSize, tick.C)
+				close(wch)
+			}()
+
+			// Wait for both our reads and writes to be finished.
+			if wErr := <-wch; wErr != nil {
+				return xerrors.Errorf("write to pty: %w", wErr)
+			}
+			if rErr := <-rch; rErr != nil {
+				return xerrors.Errorf("read from pty: %w", rErr)
+			}
+
+			duration := time.Since(start)
+
+			results := trafficGenOutput{
+				DurationSeconds: duration.Seconds(),
+				SentBytes:       crw.BytesWritten(),
+				RcvdBytes:       crw.BytesRead(),
+			}
+
+			out, err := formatter.Format(inv.Context(), results)
+			if err != nil {
+				return err
+			}
+
+			_, err = fmt.Fprintln(inv.Stdout, out)
+			return err
+		},
+	}
+
+	cmd.Options = []clibase.Option{
+		{
+			Flag:        "duration",
+			Env:         "CODER_TRAFFICGEN_DURATION",
+			Default:     "10s",
+			Description: "How long to generate traffic for.",
+			Value:       clibase.DurationOf(&duration),
+		},
+		{
+			Flag:        "bps",
+			Env:         "CODER_TRAFFICGEN_BPS",
+			Default:     "1024",
+			Description: "How much traffic to generate in bytes per second.",
+			Value:       clibase.Int64Of(&bps),
+		},
+	}
+
+	formatter.AttachOptions(&cmd.Options)
+	return cmd
+}
+
+func readContext(ctx context.Context, src io.Reader, bufSize int) error {
+	buf := make([]byte, bufSize)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			if ctx.Err() != nil {
+				return nil
+			}
+			_, err := src.Read(buf)
+			if err != nil {
+				if xerrors.Is(err, io.EOF) {
+					return nil
+				}
+				return err
+			}
+		}
+	}
+}
+
+func writeRandomData(ctx context.Context, dst io.Writer, size int, tick <-chan time.Time) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-tick:
+			payload := "#" + mustRandStr(size-1)
```

Comment:
review: the actual payload is just a commented string. This ensures we don't accidentally execute anything.